pstoreds/addr_book.go

```diff
@@ -2,398 +2,608 @@ package pstoreds
 
 import (
 	"context"
-	"encoding/binary"
-	"errors"
+	"fmt"
+	"sort"
+	"strconv"
+	"sync"
+	"sync/atomic"
 	"time"
 
-	lru "github.com/hashicorp/golang-lru"
-	base32 "github.com/whyrusleeping/base32"
-
 	ds "github.com/ipfs/go-datastore"
 	query "github.com/ipfs/go-datastore/query"
 	logging "github.com/ipfs/go-log"
 
-	ma "github.com/multiformats/go-multiaddr"
-	mh "github.com/multiformats/go-multihash"
-
 	peer "github.com/libp2p/go-libp2p-peer"
 	pstore "github.com/libp2p/go-libp2p-peerstore"
+	pb "github.com/libp2p/go-libp2p-peerstore/pb"
 	pstoremem "github.com/libp2p/go-libp2p-peerstore/pstoremem"
-)
 
-var (
-	log = logging.Logger("peerstore/ds")
-	// The maximum representable value in time.Time is time.Unix(1<<63-62135596801, 999999999).
-	// But it's too brittle and implementation-dependent, so we prefer to use 1<<62, which is in the
-	// year 146138514283. We're safe.
-	maxTime = time.Unix(1<<62, 0)
-
-	ErrTTLDatastore = errors.New("datastore must provide TTL support")
+	ma "github.com/multiformats/go-multiaddr"
+	lru "github.com/raulk/golang-lru"
+	b32 "github.com/whyrusleeping/base32"
 )
 
-// Peer addresses are stored under the following db key pattern:
-// /peers/addr/<b32 peer id no padding>/<hash of maddr>
-var abBase = ds.NewKey("/peers/addrs")
-
-var _ pstore.AddrBook = (*dsAddrBook)(nil)
-
-// dsAddrBook is an address book backed by a Datastore with both an
-// in-memory TTL manager and an in-memory address stream manager.
-type dsAddrBook struct {
-	cache        cache
-	ds           ds.TxnDatastore
-	subsManager  *pstoremem.AddrSubManager
-	writeRetries int
-}
-
 type ttlWriteMode int
 
 const (
 	ttlOverride ttlWriteMode = iota
 	ttlExtend
 )
 
-type cacheEntry struct {
-	expiration time.Time
-	addrs      []ma.Multiaddr
-}
+var (
+	log = logging.Logger("peerstore/ds")
+	// Peer addresses are stored under the following db key pattern:
+	// /peers/addrs/<b32 peer id no padding>
+	addrBookBase = ds.NewKey("/peers/addrs")
+	// GC lookahead entries are stored in keys with pattern:
+	// /peers/gc/addrs/<unix timestamp of next visit>/<peer ID b32> => nil
+	gcLookaheadBase = ds.NewKey("/peers/gc/addrs")
+	arPool          = &sync.Pool{
+		New: func() interface{} {
+			return &addrsRecord{
+				AddrBookRecord: &pb.AddrBookRecord{},
+				dirty:          false,
+			}
+		},
+	}
+)
 
-type addrRecord struct {
-	ttl  time.Duration
-	addr ma.Multiaddr
+// addrsRecord decorates the AddrBookRecord with locks and metadata.
+type addrsRecord struct {
+	sync.RWMutex
+	*pb.AddrBookRecord
+	dirty bool
 }
 
-func (ar *addrRecord) MarshalBinary() ([]byte, error) {
-	ttlB := make([]byte, 8)
-	binary.LittleEndian.PutUint64(ttlB, uint64(ar.ttl))
-	return append(ttlB, ar.addr.Bytes()...), nil
+func (r *addrsRecord) Reset() {
```

Comment:
Are these methods actually public?

(Yeah, we tend to be bad about public versus private but we might as well start getting better at it; I ask because we *have* a lock but we're not taking it here)

Reply:
I removed this method and made `clean` private.